hash_practice/exercises.py

```diff
@@ -1,29 +1,163 @@
+        anagram_array = []
+        anagram_array.append(string)
+
+        for word, char_freq_dict in grouped_string_dicts.items():
+            if grouped_string_dicts[string] == char_freq_dict and word != string:
+                anagram_array.append(word)
+                strings.remove(word)
+
+        grouped_anagrams.append(anagram_array)
+
+    return grouped_anagrams
 
 def top_k_frequent_elements(nums, k):
-    """ This method will return the k most common elements
-        In the case of a tie it will select the first occuring element.
-        Time Complexity: ?
-        Space Complexity: ?
+    """ 
+    Returns the k most common elements
+    In the case of a tie, it will select the first occuring element.
+    Time Complexity: O(n)
+    Space Complexity: O(n)
+    """
+    most_common_nums = []
+
+    if not nums:
+        return most_common_nums
+
+    freq_map = defaultdict(int)
+    for num in nums:
+        freq_map[num] += 1
+
+    while len(most_common_nums) < k:
+        most_freq_num = max(freq_map, key = lambda x: freq_map[x])
+        most_common_nums.append(most_freq_num)
+        del freq_map[most_freq_num]
+
+    return most_common_nums
+
+def valid_sudoku_rows(table):
+    """
+    Returns true if all rows in a sudoku table are valid.
+    Time Complexity: O(n^2)
+    Space Complexity: O(n)
+    """
+    for row in table:
+        row_set = set()
+
+        for char in row:
+            if char not in row_set and char != ".":
+                row_set.add(char)
+            elif char in row_set:
+                return False
+            
+    return True
+
+def valid_sudoku_columns(table):
     """
-    pass
+    Returns true if all columns in a sudoku table are valid.
+    Time Complexity: O(n)
+    Space Complexity: O(n)
+    """
+    index = 0
+
+    while index < 9: 
+        column_set = set()
+        
+        for row in table:
+            if row[index] not in column_set and row[index] != ".":
+                column_set.add(row[index])
+            
+            elif row[index] in column_set:
+                return False
+
+        index += 1
 
+    return True
+
+def valid_sudoku_box(table, left_index, right_index, top_index, bottom_index):
+    """
+    Returns true if 3x3 box, defined by indices in sudoku table, are valid.
+    Time Complexity: O(n^2)
+    Space Complexity: O(n)
+    """
+    box_set = set()
+
+    for column in range(left_index, right_index + 1):
+        for row in range (top_index, bottom_index + 1):
+            if table[column][row] not in box_set and table[column][row] != ".":
+                box_set.add(table[column][row])
+            elif table[column][row] in box_set:
+                return False
+
+    return True
+
+def valid_sudoku_boxes(table):
+    """
+    Returns true if all 3x3 boxes in a sudoku table are valid.
+    Time Complexity: O()
+    Space Complexity: O()
+    """
+
+    # I'm positive I could do this recursively, but my bran is too tired

```

Comment:
This is honestly fine, the instructor solution actually does something similar with hardcoded indices. :3